docs/writing_tests.md

```diff
@@ -0,0 +1,156 @@
+
+
+- `TEST_CASE`: This is the most basic macro, useful for test cases that do not require multiple compile-time definitions.
+- `TEMPLATE_TEST_CASE`
+- `TEMPLATE_TEST_CASE_SIG`
+
+**TODO**
+
+### Sections
+
+**TODO**
+
+### Generators
+
+**TODO**
+
+## Special Macros & Custom Matchers
+
+**TODO**
+
+## Best Practices
+
+Here is a list of best practices for writing test cases.
+These are not set in stone and are likely to evolve over time.
+
+- Always write tests using Catch2 macros, avoid [legacy test cases](#new-style-vs-legacy-test-cases).
+- Avoid old-style `if(!condition) FAIL("reason");` pattern.
+  Use `CHECK(condition)` instead.
+- Keep test cases small and focused to a single concept / behavior.
+  Even a single function could be tested with several test cases.
+- Use natural language descriptions for test cases.
+    - Avoid: `"host_accessor range mismatch exception"`.
+    - Prefer: `"host_accessors throws if accessed range exceeds buffer dimensions"`.
+- Tag test cases according to the feature being tested:
+    - Use `[some_type]` for types that exist in the SYCL specification (example: `[host_accessor]`).
+    - Use `[some-concept]` for concepts without a clearly associated type (example: `[backend-interop]`).
+- Try to write test cases in the same order as the associated API specification (if possible).
+
+## New-style vs Legacy Test Cases
+
+When browsing the CTS, you will likely encounter two different kinds of test cases: **New-style test cases** and **legacy test cases**.
+New-style test cases are written using free-standing [Catch2](https://github.com/catchorg/Catch2/) macros such as `TEST_CASE` and will look something like this:
+
+```c++
+TEST_CASE("SYCL feature XY works as expected", "[feature-xy]") {
+    // ...
+    CHECK(works_as_expected);
+}
+```
+
+Importantly, multiple of these test cases will typically be grouped into a single file.
+
+Legacy test cases on the other hand use a class-based approach, where a test case is implemented by extending the `sycl_test::util::test_base` class.
+Testing logic is then implemented in the `run` member function:
+
+```c++
+#define TEST_NAME feature_xy
+
+namespace TEST_NAMESPACE {
+using namespace sycl_cts;
+
+class TEST_NAME : public util::test_base {
+public:
+  void get_info(test_base::info &out) const override { /* ... */ }
+
+  void run(util::logger &log) override {
+      // ...
+      if(!works_as_expected) {
+          FAIL("feature XY does not work as expected");
+      }
+  }
+};
+
+util::test_proxy<TEST_NAME> proxy;
+}
+```
+
+While legacy test cases are still mapped to Catch2 under the hood, they require a lot of boilerplate code and therefore testing logic for distinct aspects of a feature are often grouped into a single test case, making them harder to comprehend and debug.
+Although technically not required, usually only one class extending `test_base` is defined per file.
```

Comment:
I'm not sure what your plans are for documenting the legacy style.  I would suggest that you do not spend much time documenting it, though.  We want people to use the Catch2 style, not the legacy one.

Reply:
This already is about the extent to which I wanted to explain legacy test cases. The only reason I'm including this at all is so that people aren't confused when they browse existing tests!